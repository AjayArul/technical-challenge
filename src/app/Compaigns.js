import React, { Component } from 'react';
import dateFormat from 'dateformat'
import ListTable from './../components/container/listTable/ListTable'
import { DATA_HEADER, DATA_LIST, DATE_FORMAT } from './../Constants'
import { formatCash } from './../Utils'

class Components  extends Component{

  constructor () {
    super()
    this.state = {
      dataList: DATA_LIST
    }
    window.AddCampaigns = this.AddCampaigns.bind(this);
  }

  // when `window.AddCampaigns("with array")` run from the browser inspect this method will call
  AddCampaigns = (list) => {
    if (!Array.isArray(list)) {
      console.log("Not an array!")
    } else {
      this.setState(prevState => ({
        dataList: [...prevState.dataList, ...list]
      }), () => {
        this.modifyData()
      })
    }
  }

  // data will format. eg: currency format and add status
  modifyData = () => {
    let dataList = this.state.dataList
    const today = new Date();
    const todayDate = dateFormat(today, DATE_FORMAT)
    dataList.map(data => {
      const startDate = dateFormat(data.startDate, DATE_FORMAT)
      const endDate = dateFormat(data.endDate, DATE_FORMAT)
      const isActive = (startDate <= todayDate && endDate >= todayDate)
      data.budget && (data.budget = formatCash(data.budget))
      data.status = <span className={`table-status text-white ${isActive ? 'bg-success' : 'bg-danger'}`}> {isActive ? 'Active' : 'Inactive'} </span>
    })

    this.setState({
      dataList: dataList
    })
  }

  componentDidMount() {
    this.modifyData()
  }

  render() {
    return (
      <div className="container mb-5">
        <h2 className='mt-3'> Campaigns </h2>
        <ListTable dataList={this.state.dataList} cols={DATA_HEADER}/>
      </div>
    )
  }
}

export default Components
