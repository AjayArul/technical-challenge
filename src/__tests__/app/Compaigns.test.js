import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Compaigns from './../../app/Compaigns';
import { DATA_LIST } from './../../Constants'

describe('Compaigns test <Compaigns />', () => {
  let _props, _wrapper, _mountwrapper
  beforeEach(() => {  
    _props = {
      dataList: [],
      AddCampaigns: jest.fn()
    }
    _wrapper = shallow(<Compaigns  {..._props} />)
    _mountwrapper = mount(<Compaigns {..._props}/>)
  })
  
  it('Should Defined and render <Compaigns /> component', () => {
      expect(_wrapper).toBeDefined()
      expect(_wrapper).toMatchSnapshot()
  });

  it('call `AddCampaigns` method, when the argument is an array', () => {
    _mountwrapper.instance().AddCampaigns = jest.fn()
    _mountwrapper.instance().AddCampaigns(DATA_LIST)
    expect(_mountwrapper.instance().AddCampaigns).toHaveBeenCalled()
  })

  it('does not call `AddCampaigns` method, when the argument is not an array', () => {
    _mountwrapper.instance().AddCampaigns(null)
    expect(_props.AddCampaigns).toHaveBeenCalledTimes(0)
  })

});