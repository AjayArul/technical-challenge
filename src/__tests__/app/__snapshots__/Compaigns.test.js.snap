// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compaigns test <Compaigns /> Should Defined and render <Compaigns /> component 1`] = `ShallowWrapper {}`;
